tracing-subscriber/src/fmt/fmt_layer.rs

```diff
@@ -0,0 +1,457 @@
+            _inner: self._inner,
+        }
+    }
+
+    /// Sets the subscriber being built to use a JSON formatter.
+    ///
+    /// See [`format::Json`]
+    #[cfg(feature = "json")]
+    pub fn json(self) -> Builder<S, format::JsonFields, format::Format<format::Json, T>, W> {
+        Builder {
+            fmt_event: self.fmt_event.json(),
+            fmt_fields: format::JsonFields::new(),
+            make_writer: self.make_writer,
+            _inner: self._inner,
+        }
+    }
+}
+
+impl<S, N, E, W> Builder<S, N, E, W> {
+    /// Sets the Visitor that the subscriber being built will use to record
+    /// fields.
+    pub fn fmt_fields<N2>(self, fmt_fields: N2) -> Builder<S, N2, E, W>
+    where
+        N2: for<'writer> FormatFields<'writer> + 'static,
+    {
+        Builder {
+            fmt_event: self.fmt_event,
+            fmt_fields,
+            make_writer: self.make_writer,
+            _inner: self._inner,
+        }
+    }
+}
+
+impl<S, N, E, W> Builder<S, N, E, W>
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+    N: for<'writer> FormatFields<'writer> + 'static,
+    E: FormatEvent<S, N> + 'static,
+    W: MakeWriter + 'static,
+{
+    /// Builds a [FmtLayer] infalliably.
+    pub fn finish(self) -> FmtLayer<S, N, E, W> {
+        FmtLayer {
+            make_writer: self.make_writer,
+            fmt_fields: self.fmt_fields,
+            fmt_event: self.fmt_event,
+            _inner: self._inner,
+        }
+    }
+}
+
+impl<S> Default for FmtLayer<S>
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+{
+    fn default() -> Self {
+        Builder::default().finish()
+    }
+}
+
+impl<S> Default for Builder<S> {
+    fn default() -> Self {
+        Self {
+            fmt_fields: format::DefaultFields::default(),
+            fmt_event: format::Format::default(),
+            make_writer: io::stdout,
+            _inner: PhantomData,
+        }
+    }
+}
+
+impl<S, N, E, W> FmtLayer<S, N, E, W>
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+    N: for<'writer> FormatFields<'writer> + 'static,
+    E: FormatEvent<S, N> + 'static,
+    W: MakeWriter + 'static,
+{
+    #[inline]
+    fn make_ctx<'a>(&'a self, ctx: Context<'a, S>) -> FmtContext<'a, S, N> {
+        FmtContext {
+            ctx,
+            fmt_fields: &self.fmt_fields,
+        }
+    }
+}
+
+/// A formatted representation of a span's fields stored in its [extensions].
+///
+/// Because `FormattedFields` is generic over the type of the formatter
+/// that produced it, multiple versions of a span's formatted fields can be
+/// stored in the [`Extensions`][extensions] type-map. This means that when
+/// multiple formatters are in use, each can store its own formatted
+/// representation without conflicting.
+///
+/// [extensions]: ../registry/extensions/index.html
+#[derive(Default)]
+pub struct FormattedFields<E> {
+    _format_event: PhantomData<fn(E)>,
+    /// The formatted fields of a span.
+    pub fmt_fields: String,
```

Comment:
maybe put this first?

Reply:
also, not convinced about this field name, and if we're making it public we should probably give it a name that we're willing to commit to...maybe just call it 
```suggestion
    pub fields: String,
```

Reply:
How is this `fields` (plural) when the type is `String`?

Reply:
I've:
- Added `ops::Deref` and `fmt::Display` implementations on `FormattedFields`. Formatting goes through the `fmt::Display` impl.
- Renamed `fmt_fields` to `fields`; made `fields` private.